### PR TITLE
Calendar Card: New Card

### PR DESCRIPTION
**Pull request by zsarnett**
## Proposed change

04786c95963d4551441af48377b6ef06

## Type of change

- [x] New feature (thank you!)

## Example configuration

```yaml
entities:
  - calendar.calendar_1
  - calendar.calendar_2
type: calendar
```

## TODO

- [ ] Link to the documentation pull request:
- [x] Calendar Card Editor
- [x] Only set eventDisplay to `list-item` when Narrow
- [ ] Add `shouldUpdate`
- [x] Make sure changes don't affect Calendar Panel (I went back to dev backend. Need to point to my dev)
- [ ] Find out why colors aren't working

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/panels/calendar/ha-full-calendar.ts [hunk 1]
```diff
@@ -34,33 +39,43 @@ declare global {
   }
 }
 
-const fullCalendarConfig = {
+const defaultFullCalendarConfig = {
   headerToolbar: false,
-  plugins: [dayGridPlugin],
+  plugins: [dayGridPlugin, listPlugin, interactionPlugin],
```
Comment (zsarnett) [inline src/panels/calendar/ha-full-calendar.ts:44]: I would like to only load the code for the plugins we need based on the options given to the Calendar Element. Not sure the best way to do that as well as if I should put that in a new PR

File: src/panels/calendar/ha-full-calendar.ts [hunk 2]
```diff
@@ -159,7 +174,24 @@ class HAFullCalendar extends LitElement {
   }
 
   protected firstUpdated(): void {
-    const config = { ...fullCalendarConfig, locale: this.hass.language };
+    const config = {
+      ...defaultFullCalendarConfig,
+      locale: this.hass.language,
+      views: {
+        dayGridMonth: {
+          eventDisplay: "list-item",
```
Comment (zsarnett) [inline src/panels/calendar/ha-full-calendar.ts:182]: Only do this when narrow...

File: src/panels/lovelace/cards/hui-calendar-card.ts [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+import "../components/hui-warning";
+import { processConfigEntities } from "../common/process-config-entities";
+import { computeStateName } from "../../../common/entity/compute_state_name";
+import { fetchCalendarEvents } from "../../../data/calendar";
+import { HASSDomEvent } from "../../../common/dom/fire_event";
+import { HA_COLOR_PALETTE } from "../../../common/const";
+
+@customElement("hui-calendar-card")
+export class HuiCalendarCard extends LitElement implements LovelaceCard {
+  // public static async getConfigElement(): Promise<LovelaceCardEditor> {
```
Comment (zsarnett) [inline src/panels/lovelace/cards/hui-calendar-card.ts:35]: TODO

File: src/panels/lovelace/cards/hui-calendar-card.ts [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+      return html``;
+    }
+
+    return html`
+      <ha-card .header=${this._config.title}>
+        <ha-full-calendar
+          .events=${this._events}
+          .narrow=${true}
+          .hass=${this.hass}
+          .views=${["listWeek", "dayGridMonth"]}
```
Comment (zsarnett) [inline src/panels/lovelace/cards/hui-calendar-card.ts:97]: This tells me `string[] is not assignable to FullCalendarView[]`

File: src/panels/calendar/ha-full-calendar.ts (hunk 1 above)
Comment (zsarnett) [reply src/panels/calendar/ha-full-calendar.ts:44]: I have an Idea for this. Using the same methods for edit code in the view but would rather save it for the PR after this one if possible.

**Comment by zsarnett** [comment]
image

**Comment by zsarnett** [comment]
I can't figure out why Colors aren't working in the card but work on the Panel. Even if the panel is in Narrow mode it still works on the panel...

**Comment by zsarnett** [comment]
Probably need to add resize observer

**Comment by zsarnett** [comment]
I feel like there is one more issue but I cant remember what it is right now and I cant find it by playing with the card.... Just remember I needed to fix something

**Comment by bramkragten** [comment]
Playing with this a few things I noticed:

- Clicking on an event doesn't open day view
- Day view in missing in select view options, but you can get there by clicking a day on the month view, feels weird
- I miss padding on the card, I understand it can get too small then, but I think should solve that with restricting views to sizes with a resizeObsever
- It is missing from the card picker

File: src/components/ha-button-toggle-group.ts [hunk 5]
```diff
@@ -11,26 +11,28 @@ import {
 import "./ha-icon-button";
 
 import { fireEvent } from "../common/dom/fire_event";
-import type { ToggleButton } from "../types";
+import type { ToggleViewButton } from "../types";
```
Comment (bramkragten) [inline src/components/ha-button-toggle-group.ts:14]: Why was this changed to a less generic type?

Comment (zsarnett) [reply src/components/ha-button-toggle-group.ts:14]: Because it should be a less generic type right? These buttons are only view buttons and the value can only be a Full Calendar view

**Comment by zsarnett** [comment]
I will be able to take a look at this more next week or the week after. Getting to TX on June 1st :)

File: src/panels/lovelace/editor/config-elements/hui-calendar-card-editor.ts [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+  private _valueChanged(ev: EntitiesEditorEvent | CustomEvent): void {
+    if (!this._config || !this.hass) {
+      return;
+    }
+
+    const target = ev.target! as EditorTarget;
+
+    if (
+      (target.configValue! === "title" && target.value === this._title) ||
+      (target.configValue! === "theme" && target.value === this._theme)
```
Comment (bramkragten) [inline src/panels/lovelace/editor/config-elements/hui-calendar-card-editor.ts:106]:
```suggestion
      target.value === this.[`_${target.configValue}`]
```

File: src/panels/lovelace/editor/config-elements/hui-calendar-card-editor.ts [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+      this._config.entities = ev.detail.value;
+      this._configEntities = processEditorEntities(this._config.entities);
+    } else if (target.configValue) {
+      if (target.value === "") {
+        delete this._config[target.configValue!];
+      } else {
+        this._config = {
+          ...this._config,
+          [target.configValue]:
+            target.checked !== undefined ? target.checked : target.value,
```
Comment (bramkragten) [inline src/panels/lovelace/editor/config-elements/hui-calendar-card-editor.ts:121]: You don't have any checkboxes?

Comment (zsarnett) [reply src/panels/lovelace/editor/config-elements/hui-calendar-card-editor.ts:121]: Copy pasta is a bitch

File: src/panels/lovelace/editor/config-elements/hui-calendar-card-editor.ts [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    if (
+      (target.configValue! === "title" && target.value === this._title) ||
+      (target.configValue! === "theme" && target.value === this._theme)
+    ) {
+      return;
+    }
+
+    if (ev.detail && ev.detail.value && Array.isArray(ev.detail.value)) {
+      this._config.entities = ev.detail.value;
+      this._configEntities = processEditorEntities(this._config.entities);
```
Comment (bramkragten) [inline src/panels/lovelace/editor/config-elements/hui-calendar-card-editor.ts:113]: `processEditorEntities` should be used when the entities should be objects, we can just use an array here with entityIds right?

File: src/panels/lovelace/cards/hui-calendar-card.ts [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+        ></ha-full-calendar>
+      </ha-card>
+    `;
+  }
+
+  protected firstUpdated(changedProps: PropertyValues): void {
+    super.firstUpdated(changedProps);
+    const configEntities = this._config!.entities
+      ? processConfigEntities(this._config!.entities)
+      : [];
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-calendar-card.ts:109]: Move this to `setConfig`

File: src/panels/lovelace/cards/hui-calendar-card.ts [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+
+    configEntities.forEach((entity, idx) => {
+      calendars.push({
+        entity_id: entity.entity,
+        name: computeStateName(this.hass!.states[entity.entity]),
+        backgroundColor: HA_COLOR_PALETTE[idx % HA_COLOR_PALETTE.length],
+      });
+    });
+
+    this._calendars = calendars;
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-calendar-card.ts:121]: This should also go to `setConfig`, if you change something now it is not reflected in the card.

File: src/panels/lovelace/cards/hui-calendar-card.ts [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+    const oldHass = changedProps.get("hass") as HomeAssistant | undefined;
+    const oldConfig = changedProps.get("_config") as
+      | CalendarCardConfig
+      | undefined;
+
+    if (
+      !oldHass ||
+      !oldConfig ||
+      oldHass.themes !== this.hass.themes ||
+      oldConfig.theme !== this._config.theme
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-calendar-card.ts:139]:
```suggestion
      changedProps.hass("hass") && oldHass.themes !== this.hass.themes ||
      changedProps.hass("_config") && oldConfig.theme !== this._config.theme
```

File: src/panels/lovelace/cards/hui-calendar-card.ts (hunk 4 above)
Comment (bramkragten) [reply src/panels/lovelace/cards/hui-calendar-card.ts:98]: You can ignore it. Or try `["listWeek", "dayGridMonth"] as FullCalendarView[]`

File: src/panels/lovelace/cards/hui-calendar-card.ts [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+      ? processConfigEntities(this._config!.entities)
+      : [];
+
+    const calendars: Calendar[] = [];
+
+    configEntities.forEach((entity, idx) => {
+      calendars.push({
+        entity_id: entity.entity,
+        name: computeStateName(this.hass!.states[entity.entity]),
+        backgroundColor: HA_COLOR_PALETTE[idx % HA_COLOR_PALETTE.length],
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-calendar-card.ts:117]:
```suggestion
      backgroundColor: `#${HA_COLOR_PALETTE[idx % HA_COLOR_PALETTE.length]}`,
```

Comment (zsarnett) [reply src/panels/lovelace/cards/hui-calendar-card.ts:117]: HOLY 
Thank oyu

File: src/components/ha-button-toggle-group.ts (hunk 5 above)
Comment (bramkragten) [reply src/components/ha-button-toggle-group.ts:14]: Why? It is for a button toggle group element. It doesn't have to toggle a view.

Comment (zsarnett) [reply src/components/ha-button-toggle-group.ts:14]: I think I was drunk when I did this. I honestly don't know what I was thinking

**Comment by zsarnett** [comment]
@bramkragten  I need some help. I have somehow broken the calls to the calendar API that if I load a view with the Calendar Card on it, it freezes the tab...

I have look at this many times and I am missing something that I am looking over. If you find time to look this over I would appreciate it. This is long over due and I am trying to find the time to get this back into my day

File: src/panels/lovelace/cards/hui-calendar-card.ts [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    const calendars: Calendar[] = [];
+
+    config!.entities.forEach((entity, idx) => {
+      calendars.push({
+        entity_id: entity,
+        backgroundColor: `#${HA_COLOR_PALETTE[idx % HA_COLOR_PALETTE.length]}`,
+      });
+    });
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-calendar-card.ts:85]:
```suggestion
    this._calendars = config!.entities.map((entity, idx) => {
      return {
        entity_id: entity,
        backgroundColor: `#${HA_COLOR_PALETTE[idx % HA_COLOR_PALETTE.length]}`,
      };
    });
```

File: src/panels/lovelace/cards/hui-calendar-card.ts [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+    };
+  }
+
+  @property() public hass?: HomeAssistant;
+
+  @property() private _config?: CalendarCardConfig;
+
+  @property() public _events: CalendarEvent[] = [];
+
+  @property() private _calendars: Calendar[] = [];
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-calendar-card.ts:67]: Use `internalProperty` for private properties.
```suggestion
  @internalProperty() private _calendars: Calendar[] = [];
```

File: src/panels/lovelace/cards/hui-calendar-card.ts [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+    if (!this._config || !this.hass || !this._calendars.length) {
+      return html``;
+    }
+
+    return html`
+      <ha-card>
+        <div class="header">${this._config.title}</div>
+        <ha-full-calendar
+          .events=${this._events}
+          .narrow=${this.offsetWidth < 870}
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-calendar-card.ts:106]: You want to do this with a `resizeObserver`

File: src/panels/calendar/ha-full-calendar.ts [hunk 16]
```diff
@@ -60,12 +60,11 @@ const viewButtons: ToggleButton[] = [
 class HAFullCalendar extends LitElement {
   public hass!: HomeAssistant;
 
-  @property({ type: Boolean, reflect: true })
-  public narrow!: boolean;
+  @property({ type: Boolean, reflect: true }) public narrow!: boolean;
```
Comment (zsarnett) [inline src/panels/calendar/ha-full-calendar.ts:63]: This breaks the card for some reason. When Reflect is set to true on this property and then in the Card, if you set `.narrow=${true}` it freezes the browser... Not sure why this is happening @bramkragten

Comment (bramkragten) [reply src/panels/calendar/ha-full-calendar.ts:63]: You should be able to just set the attribute `narrow`, is this value always set? Or should you initialize it to false?

Comment (zsarnett) [reply src/panels/calendar/ha-full-calendar.ts:63]: It should probably be always set by the user of the element right. They set narrow and if not true then false. Should never be a null for a boolean

Comment (bramkragten) [reply src/panels/calendar/ha-full-calendar.ts]: true

File: src/components/ha-button-toggle-group.ts [hunk 17]
```diff
@@ -15,22 +15,28 @@ import type { ToggleButton } from "../types";
...
-          (button) => html` <ha-icon-button
-            .label=${button.label}
-            .icon=${button.icon}
-            .value=${button.value}
-            ?active=${this.active === button.value}
-            @click=${this._handleClick}
-          >
-          </ha-icon-button>`
+          (button) => html`
+            <ha-icon-button
```
Comment (bramkragten) [inline src/components/ha-button-toggle-group.ts:31]: We should avoid the use of `ha-icon-button` and `hass:icon` and use `ha-svg-icon` if the icon is not user configurable.

File: src/components/ha-button-toggle-group.ts [hunk 18]
```diff
@@ -15,22 +15,28 @@ import type { ToggleButton } from "../types";
...
 export class HaButtonToggleGroup extends LitElement {
-  @property() public buttons!: ToggleButton[];
+  @property({ attribute: false }) public buttons!: ToggleButton[];
 
   @property() public active?: string;
 
   protected render(): TemplateResult {
+    if (this.buttons.length <= 1) {
+      return html``;
+    }
```
Comment (bramkragten) [inline src/components/ha-button-toggle-group.ts:25]: Why don't we render a single button?

File: src/panels/calendar/ha-full-calendar.ts [hunk 19]
```diff
@@ -25,43 +30,58 @@ import type {
...
-  plugins: [dayGridPlugin],
+  plugins: [dayGridPlugin, listPlugin, interactionPlugin],
   initialView: "dayGridMonth",
   dayMaxEventRows: true,
   height: "parent",
+  eventDisplay: "list-item",
 };
 
 const viewButtons: ToggleButton[] = [
   { label: "Month View", value: "dayGridMonth", icon: "hass:view-module" },
```
Comment (bramkragten) [inline src/panels/calendar/ha-full-calendar.ts:57]:
```suggestion
  { label: "Month View", value: "dayGridMonth", icon: mdiViewModule },
```

File: src/panels/calendar/ha-full-calendar.ts [hunk 20]
```diff
@@ -25,43 +30,58 @@ import type {
...
 
   @internalProperty() private calendar?: Calendar;
 
-  @internalProperty() private _activeView = "dayGridMonth";
+  @internalProperty() private _activeView: FullCalendarView = "dayGridMonth";
 
   protected render(): TemplateResult {
+    const viewToggleButtons = viewButtons.filter((button) =>
+      this.views.includes(button.value as FullCalendarView)
+    );
```
Comment (bramkragten) [inline src/panels/calendar/ha-full-calendar.ts:83]: Use memoize for this

File: src/panels/calendar/ha-full-calendar.ts [hunk 21]
```diff
@@ -157,10 +177,37 @@ class HAFullCalendar extends LitElement {
...
+
+    config.eventClick = (info) => {
+      if (info.view.type !== "dayGridMonth") {
+        return;
+      }
+
+      this._activeView = "dayGridDay";
+      this.calendar!.changeView("dayGridDay");
+      this.calendar!.gotoDate(info.event.startStr);
+    };
```
Comment (bramkragten) [inline src/panels/calendar/ha-full-calendar.ts:210]: Can we make these class functions?
```
config.eventClick = this._handleEventClick;
```

**Review by bramkragten — APPROVED** [review]

File: src/components/ha-button-toggle-group.ts [hunk 22]
```diff
@@ -7,35 +7,33 @@ import {
...
   css,
 } from "lit-element";
-
-import "./ha-icon-button";
+import { classMap } from "lit-html/directives/class-map";
 
 import { fireEvent } from "../common/dom/fire_event";
 import type { ToggleButton } from "../types";
 
+import "./ha-icon-button";
```
Comment (zsarnett) [inline src/components/ha-button-toggle-group.ts:15]:
```suggestion
```

File: src/components/ha-button-toggle-group.ts [hunk 23]
```diff
@@ -7,30 +7,33 @@ import {
...
-          (button) => html` <ha-icon-button
-            .label=${button.label}
-            .icon=${button.icon}
-            .value=${button.value}
-            ?active=${this.active === button.value}
-            @click=${this._handleClick}
-          >
-          </ha-icon-button>`
+          (button) => html`
+            <ha-svg-icon
```
Comment (bramkragten) [inline src/components/ha-button-toggle-group.ts:28]: Should these not be wrapped in a `<mwc-icon-button>`?

Comment (zsarnett) [reply src/components/ha-button-toggle-group.ts:28]: Probably. I was confused on exactly what to do when not using the ha-button. but that makes more sense 😅

**Review by bramkragten — APPROVED** [review]
